hpy/devel/src/runtime/ctx_type.c

```diff
@@ -0,0 +1,212 @@
+        count = 0;
+    }
+    else {
+        count = 0;
+        while (hpymethods[count].ml_name != NULL)
+            count++;
+    }
+
+    // allocate&fill the result
+    PyMethodDef *result = PyMem_Malloc(sizeof(PyMethodDef) * (count+1));
+    if (result == NULL) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+    for(int i=0; i<count; i++) {
+        HPyMethodDef *src = &hpymethods[i];
+        PyMethodDef *dst = &result[i];
+        dst->ml_name = src->ml_name;
+        dst->ml_doc = src->ml_doc;
+        methoddef_copy_ml_meth_and_flags(src, dst);
+    }
+    result[count] = (PyMethodDef){NULL, NULL, 0, NULL};
+    return result;
+}
+
+static PyType_Slot *
+create_slot_defs(HPyType_Spec *hpyspec)
+{
+    // count the slots
+    Py_ssize_t count;
+    if (hpyspec->slots == NULL) {
+        count = 0;
+    }
+    else {
+        count = 0;
+        while (hpyspec->slots[count].slot != 0)
+            count++;
+    }
+
+    // allocate&fill the result
+    PyType_Slot *result = PyMem_Malloc(sizeof(PyType_Slot) * (count+1));
+    if (result == NULL)
+        return NULL;
+    for(int i=0; i<count; i++) {
+        HPyType_Slot *src = &hpyspec->slots[i];
+        PyType_Slot *dst = &result[i];
+        dst->slot = src->slot;
+        if (src->slot == Py_tp_methods) {
+            // src->pfunc contains a HPyMethodDef*, which we convert to the
+            // CPython's equivalent. The result of create_method_defs can't be
+            // freed because CPython stores an internal reference to it, so
+            // this creates a small leak :(
+            HPyMethodDef *hpymethods = (HPyMethodDef*)src->pfunc;
+            dst->pfunc = create_method_defs(hpymethods);
+            if (dst->pfunc == NULL) {
+                PyMem_Free(result);
+                return NULL;
+            }
+        }
+        else {
+            slot_copy_pfunc(&hpyspec->slots[i], dst);
+        }
+    }
+    result[count] = (PyType_Slot){0, NULL};
+    return result;
+}
+
+_HPy_HIDDEN HPy
+ctx_Type_FromSpec(HPyContext ctx, HPyType_Spec *hpyspec)
+{
+    PyType_Spec *spec = PyMem_Malloc(sizeof(PyType_Spec));
+    if (spec == NULL) {
+        PyErr_NoMemory();
+        return HPy_NULL;
+    }
+    if (hpyspec->has_pyobject_head) {
+        PyErr_SetString(PyExc_NotImplementedError, "has_pyobject_head not supported yet");
+        return HPy_NULL;
+    }
+    spec->name = hpyspec->name;
+    spec->basicsize = hpyspec->basicsize + PyObject_HEAD_SIZE;
+    spec->itemsize = hpyspec->itemsize;
+    spec->flags = hpyspec->flags;
+    spec->slots = create_slot_defs(hpyspec);
+    if (spec->slots == NULL) {
+        PyMem_Free(spec);
+        PyErr_NoMemory();
+        return HPy_NULL;
+    }
+    PyObject *result = PyType_FromSpec(spec);
+    /* note that we do NOT free the memory which was allocated by
+       create_method_defs, because that one is referenced internally by
+       CPython (which probably assumes it's statically allocated) */
```

Comment:
We should probably come up with a general way to attach extra C-level information to type objects created by HPy.  There doesn't seem to be an official way.  We can find hacks though...

Reply:
Well, for this specific case I claim that it's CPython's fault, but I agree about the general idea. Are you ok to merge this PR now and think about this "eventually"? :)

Reply:
Resolve the issue and merging it anyway, we can always fix it later